BUG: `interpolate/tests/test_interpnd.py::TestLinearNDInterpolation::test_threading` failure

### Describe your issue.

We're getting a crash during `interpolate/tests/test_interpnd.py::TestLinearNDInterpolation::test_threading` on "Oldest GCC & pydata/sparse, fast, py3.10/npMin, pip+pytest". 

### Error message

```shell
=================================== FAILURES ===================================
______________________ interpolate/tests/test_interpnd.py ______________________
[gw0] linux -- Python 3.10.15 /opt/hostedtoolcache/Python/3.10.15/x64/bin/python
worker 'gw0' crashed while running 'interpolate/tests/test_interpnd.py::TestLinearNDInterpolation::test_threading'
```

<details>

~~`io/tests/test_mmio.py::test_threadpoolctl[scipy.io._fast_matrix_market]` is failing in the "free-threaded" and "Prerelease deps & coverage report, full, py3.10/npMin & py3.11/npPre, dev.py (3.11)" jobs (e.g. gh-21883).~~ This failure has disappeared.

```
_______________ test_threadpoolctl[scipy.io._fast_matrix_market] _______________
[gw1] linux -- Python 3.11.10 /opt/hostedtoolcache/Python/3.11.10/x64/bin/python
scipy/io/tests/test_mmio.py:811: in test_threadpoolctl
    assert_equal(fmm.PARALLELISM, 4)
E   AssertionError: 
E   Items are not equal:
E    ACTUAL: 0
E    DESIRED: 4
        threadpoolctl = <module 'threadpoolctl' from '/opt/hostedtoolcache/Python/3.11.10/x64/lib/python3.11/site-packages/threadpoolctl.py'>
```

</details>


### SciPy/NumPy/Python version and system information

<details>

```shell
Run python -m pip install cython pythran ninja meson-python pybind11 click rich_click pydevtool
Collecting cython
  Downloading Cython-3.0.11-cp311-cp311-manylinux_2_17_x86_64.manylinux2014_x86_64.whl.metadata (3.2 kB)
Collecting pythran
  Downloading pythran-0.17.0-py3-none-any.whl.metadata (7.9 kB)
Collecting ninja
  Downloading ninja-1.11.1.1-py2.py3-none-manylinux1_x86_64.manylinux_2_5_x86_64.whl.metadata (5.3 kB)
Collecting meson-python
  Downloading meson_python-0.17.1-py3-none-any.whl.metadata (4.1 kB)
Collecting pybind11
  Downloading pybind11-2.13.6-py3-none-any.whl.metadata (9.5 kB)
Collecting click
  Downloading click-8.1.7-py3-none-any.whl.metadata (3.0 kB)
Collecting rich_click
  Downloading rich_click-1.8.3-py3-none-any.whl.metadata (7.9 kB)
Collecting pydevtool
  Downloading pydevtool-0.3.0-py3-none-any.whl.metadata (934 bytes)
Collecting ply>=3.4 (from pythran)
  Downloading ply-3.11-py2.py3-none-any.whl.metadata (844 bytes)
Requirement already satisfied: setuptools in /opt/hostedtoolcache/Python/3.11.10/x64/lib/python3.11/site-packages (from pythran) (65.5.0)
Collecting gast~=0.6.0 (from pythran)
  Downloading gast-0.6.0-py3-none-any.whl.metadata (1.3 kB)
Collecting numpy (from pythran)
  Downloading numpy-2.1.3-cp311-cp311-manylinux_2_17_x86_64.manylinux2014_x86_64.whl.metadata (62 kB)
Collecting beniget~=0.4.0 (from pythran)
  Downloading beniget-0.4.2.post1-py3-none-any.whl.metadata (820 bytes)
Collecting meson>=0.63.3 (from meson-python)
  Downloading meson-1.6.0-py3-none-any.whl.metadata (1.8 kB)
Collecting packaging>=19.0 (from meson-python)
  Downloading packaging-24.2-py3-none-any.whl.metadata (3.2 kB)
Collecting pyproject-metadata>=0.7.1 (from meson-python)
  Downloading pyproject_metadata-0.9.0-py3-none-any.whl.metadata (6.9 kB)
Collecting rich>=10.7 (from rich_click)
  Downloading rich-13.9.4-py3-none-any.whl.metadata (18 kB)
Collecting typing-extensions (from rich_click)
  Downloading typing_extensions-4.12.2-py3-none-any.whl.metadata (3.0 kB)
Collecting doit>=0.36.0 (from pydevtool)
  Downloading doit-0.36.0-py3-none-any.whl.metadata (4.9 kB)
Collecting cloudpickle (from doit>=0.36.0->pydevtool)
  Downloading cloudpickle-3.1.0-py3-none-any.whl.metadata (7.0 kB)
Collecting importlib-metadata>=4.4 (from doit>=0.36.0->pydevtool)
  Downloading importlib_metadata-8.5.0-py3-none-any.whl.metadata (4.8 kB)
Collecting markdown-it-py>=2.2.0 (from rich>=10.7->rich_click)
  Downloading markdown_it_py-3.0.0-py3-none-any.whl.metadata (6.9 kB)
Collecting pygments<3.0.0,>=2.13.0 (from rich>=10.7->rich_click)
  Downloading pygments-2.18.0-py3-none-any.whl.metadata (2.5 kB)
Collecting zipp>=3.20 (from importlib-metadata>=4.4->doit>=0.36.0->pydevtool)
  Downloading zipp-3.21.0-py3-none-any.whl.metadata (3.7 kB)
Collecting mdurl~=0.1 (from markdown-it-py>=2.2.0->rich>=10.7->rich_click)
  Downloading mdurl-0.1.2-py3-none-any.whl.metadata (1.6 kB)
Downloading Cython-3.0.11-cp311-cp311-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (3.6 MB)
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 3.6/3.6 MB 154.3 MB/s eta 0:00:00
Downloading pythran-0.17.0-py3-none-any.whl (4.3 MB)
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 4.3/4.3 MB 129.1 MB/s eta 0:00:00
Downloading ninja-1.11.1.1-py2.py3-none-manylinux1_x86_64.manylinux_2_5_x86_64.whl (307 kB)
Downloading meson_python-0.17.1-py3-none-any.whl (27 kB)
Downloading pybind11-2.13.6-py3-none-any.whl (243 kB)
Downloading click-8.1.7-py3-none-any.whl (97 kB)
Downloading rich_click-1.8.3-py3-none-any.whl (35 kB)
Downloading pydevtool-0.3.0-py3-none-any.whl (12 kB)
Downloading beniget-0.4.2.post1-py3-none-any.whl (17 kB)
Downloading doit-0.36.0-py3-none-any.whl (85 kB)
Downloading gast-0.6.0-py3-none-any.whl (21 kB)
Downloading meson-1.6.0-py3-none-any.whl (973 kB)
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 973.9/973.9 kB 87.8 MB/s eta 0:00:00
Downloading packaging-24.2-py3-none-any.whl (65 kB)
Downloading ply-3.11-py2.py3-none-any.whl (49 kB)
Downloading pyproject_metadata-0.9.0-py3-none-any.whl (18 kB)
Downloading rich-13.9.4-py3-none-any.whl (242 kB)
Downloading numpy-2.1.3-cp311-cp311-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (16.3 MB)
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 16.3/16.3 MB 186.4 MB/s eta 0:00:00
Downloading typing_extensions-4.12.2-py3-none-any.whl (37 kB)
Downloading importlib_metadata-8.5.0-py3-none-any.whl (26 kB)
Downloading markdown_it_py-3.0.0-py3-none-any.whl (87 kB)
Downloading pygments-2.18.0-py3-none-any.whl (1.2 MB)
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 1.2/1.2 MB 118.6 MB/s eta 0:00:00
Downloading cloudpickle-3.1.0-py3-none-any.whl (22 kB)
Downloading mdurl-0.1.2-py3-none-any.whl (10.0 kB)
Downloading zipp-3.21.0-py3-none-any.whl (9.6 kB)
Installing collected packages: ply, ninja, zipp, typing-extensions, pygments, pybind11, packaging, numpy, meson, mdurl, gast, cython, cloudpickle, click, pyproject-metadata, markdown-it-py, importlib-metadata, beniget, rich, pythran, meson-python, doit, rich_click, pydevtool
Successfully installed beniget-0.4.2.post1 click-8.1.7 cloudpickle-3.1.0 cython-3.0.11 doit-0.36.0 gast-0.6.0 importlib-metadata-8.5.0 markdown-it-py-3.0.0 mdurl-0.1.2 meson-1.6.0 meson-python-0.17.1 ninja-1.11.1.1 numpy-2.1.3 packaging-24.2 ply-3.11 pybind11-2.13.6 pydevtool-0.3.0 pygments-2.18.0 pyproject-metadata-0.9.0 pythran-0.17.0 rich-13.9.4 rich_click-1.8.3 typing-extensions-4.12.2 zipp-3.21.0
Collecting pytest
  Downloading pytest-8.3.3-py3-none-any.whl.metadata (7.5 kB)
Collecting pytest-cov
  Downloading pytest_cov-6.0.0-py3-none-any.whl.metadata (27 kB)
Collecting pytest-xdist
  Downloading pytest_xdist-3.6.1-py3-none-any.whl.metadata (4.3 kB)
Collecting mpmath
  Downloading mpmath-1.4.0a2-py3-none-any.whl.metadata (9.2 kB)
Collecting gmpy2
  Downloading gmpy2-2.2.1-cp311-cp311-manylinux_2_17_x86_64.manylinux2014_x86_64.whl.metadata (2.8 kB)
Collecting threadpoolctl
  Downloading threadpoolctl-3.5.0-py3-none-any.whl.metadata (13 kB)
Collecting pooch
  Downloading pooch-1.8.2-py3-none-any.whl.metadata (10 kB)
Collecting hypothesis
  Downloading hypothesis-6.118.8-py3-none-any.whl.metadata (6.0 kB)
Collecting matplotlib
  Downloading matplotlib-3.10.0rc1-cp311-cp311-manylinux_2_17_x86_64.manylinux2014_x86_64.whl.metadata (11 kB)
Collecting iniconfig (from pytest)
  Downloading iniconfig-2.0.0-py3-none-any.whl.metadata (2.6 kB)
Requirement already satisfied: packaging in /opt/hostedtoolcache/Python/3.11.10/x64/lib/python3.11/site-packages (from pytest) (24.2)
Collecting pluggy<2,>=1.5 (from pytest)
  Downloading pluggy-1.5.0-py3-none-any.whl.metadata (4.8 kB)
Collecting coverage>=7.5 (from coverage[toml]>=7.5->pytest-cov)
  Downloading coverage-7.6.4-cp311-cp311-manylinux_2_5_x86_64.manylinux1_x86_64.manylinux_2_17_x86_64.manylinux2014_x86_64.whl.metadata (8.2 kB)
Collecting execnet>=2.1 (from pytest-xdist)
  Downloading execnet-2.1.1-py3-none-any.whl.metadata (2.9 kB)
Collecting platformdirs>=2.5.0 (from pooch)
  Downloading platformdirs-4.3.6-py3-none-any.whl.metadata (11 kB)
Collecting requests>=2.19.0 (from pooch)
  Downloading requests-2.32.3-py3-none-any.whl.metadata (4.6 kB)
Collecting attrs>=22.2.0 (from hypothesis)
  Downloading attrs-24.2.0-py3-none-any.whl.metadata (11 kB)
Collecting sortedcontainers<3.0.0,>=2.1.0 (from hypothesis)
  Downloading sortedcontainers-2.4.0-py2.py3-none-any.whl.metadata (10 kB)
Collecting contourpy>=1.0.1 (from matplotlib)
  Downloading contourpy-1.3.1-cp311-cp311-manylinux_2_17_x86_64.manylinux2014_x86_64.whl.metadata (5.4 kB)
Collecting cycler>=0.10 (from matplotlib)
  Downloading cycler-0.12.1-py3-none-any.whl.metadata (3.8 kB)
Collecting fonttools>=4.22.0 (from matplotlib)
  Downloading fonttools-4.54.1-cp311-cp311-manylinux_2_17_x86_64.manylinux2014_x86_64.whl.metadata (163 kB)
Collecting kiwisolver>=1.3.1 (from matplotlib)
  Downloading kiwisolver-1.4.7-cp311-cp311-manylinux_2_17_x86_64.manylinux2014_x86_64.whl.metadata (6.3 kB)
Requirement already satisfied: numpy>=1.23 in /opt/hostedtoolcache/Python/3.11.10/x64/lib/python3.11/site-packages (from matplotlib) (2.1.3)
Collecting pillow>=8 (from matplotlib)
  Downloading pillow-11.0.0-cp311-cp311-manylinux_2_28_x86_64.whl.metadata (9.1 kB)
Collecting pyparsing>=2.3.1 (from matplotlib)
  Downloading pyparsing-3.2.0-py3-none-any.whl.metadata (5.0 kB)
Collecting python-dateutil>=2.7 (from matplotlib)
  Downloading python_dateutil-2.9.0.post0-py2.py3-none-any.whl.metadata (8.4 kB)
Collecting six>=1.5 (from python-dateutil>=2.7->matplotlib)
  Downloading six-1.16.0-py2.py3-none-any.whl.metadata (1.8 kB)
Collecting charset-normalizer<4,>=2 (from requests>=2.19.0->pooch)
  Downloading charset_normalizer-3.4.0-cp311-cp311-manylinux_2_17_x86_64.manylinux2014_x86_64.whl.metadata (34 kB)
Collecting idna<4,>=2.5 (from requests>=2.19.0->pooch)
  Downloading idna-3.10-py3-none-any.whl.metadata (10 kB)
Collecting urllib3<3,>=1.21.1 (from requests>=2.19.0->pooch)
  Downloading urllib3-2.2.3-py3-none-any.whl.metadata (6.5 kB)
Collecting certifi>=2017.4.17 (from requests>=2.19.0->pooch)
  Downloading certifi-2024.8.30-py3-none-any.whl.metadata (2.2 kB)
Downloading pytest-8.3.3-py3-none-any.whl (342 kB)
Downloading pytest_cov-6.0.0-py3-none-any.whl (22 kB)
Downloading pytest_xdist-3.6.1-py3-none-any.whl (46 kB)
Downloading mpmath-1.4.0a2-py3-none-any.whl (558 kB)
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 558.8/558.8 kB 46.0 MB/s eta 0:00:00
Downloading gmpy2-2.2.1-cp311-cp311-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (1.7 MB)
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 1.7/1.7 MB 117.0 MB/s eta 0:00:00
Downloading threadpoolctl-3.5.0-py3-none-any.whl (18 kB)
Downloading pooch-1.8.2-py3-none-any.whl (64 kB)
Downloading hypothesis-6.118.8-py3-none-any.whl (471 kB)
Downloading matplotlib-3.10.0rc1-cp311-cp311-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (8.6 MB)
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 8.6/8.6 MB 98.5 MB/s eta 0:00:00
Downloading attrs-24.2.0-py3-none-any.whl (63 kB)
Downloading contourpy-1.3.1-cp311-cp311-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (326 kB)
Downloading coverage-7.6.4-cp311-cp311-manylinux_2_5_x86_64.manylinux1_x86_64.manylinux_2_17_x86_64.manylinux2014_x86_64.whl (238 kB)
Downloading cycler-0.12.1-py3-none-any.whl (8.3 kB)
Downloading execnet-2.1.1-py3-none-any.whl (40 kB)
Downloading fonttools-4.54.1-cp311-cp311-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (4.9 MB)
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 4.9/4.9 MB 119.9 MB/s eta 0:00:00
Downloading kiwisolver-1.4.7-cp311-cp311-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (1.4 MB)
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 1.4/1.4 MB 129.3 MB/s eta 0:00:00
Downloading pillow-11.0.0-cp311-cp311-manylinux_2_28_x86_64.whl (4.4 MB)
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 4.4/4.4 MB 155.3 MB/s eta 0:00:00
Downloading platformdirs-4.3.6-py3-none-any.whl (18 kB)
Downloading pluggy-1.5.0-py3-none-any.whl (20 kB)
Downloading pyparsing-3.2.0-py3-none-any.whl (106 kB)
Downloading python_dateutil-2.9.0.post0-py2.py3-none-any.whl (229 kB)
Downloading requests-2.32.3-py3-none-any.whl (64 kB)
Downloading sortedcontainers-2.4.0-py2.py3-none-any.whl (29 kB)
Downloading iniconfig-2.0.0-py3-none-any.whl (5.9 kB)
Downloading certifi-2024.8.30-py3-none-any.whl (167 kB)
Downloading charset_normalizer-3.4.0-cp311-cp311-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (142 kB)
Downloading idna-3.10-py3-none-any.whl (70 kB)
Downloading six-1.16.0-py2.py3-none-any.whl (11 kB)
Downloading urllib3-2.2.3-py3-none-any.whl (126 kB)
Installing collected packages: sortedcontainers, urllib3, threadpoolctl, six, pyparsing, pluggy, platformdirs, pillow, mpmath, kiwisolver, iniconfig, idna, gmpy2, fonttools, execnet, cycler, coverage, contourpy, charset-normalizer, certifi, attrs, requests, python-dateutil, pytest, hypothesis, pytest-xdist, pytest-cov, pooch, matplotlib
Successfully installed attrs-24.2.0 certifi-2024.8.30 charset-normalizer-3.4.0 contourpy-1.3.1 coverage-7.6.4 cycler-0.12.1 execnet-2.1.1 fonttools-4.54.1 gmpy2-2.2.1 hypothesis-6.118.8 idna-3.10 iniconfig-2.0.0 kiwisolver-1.4.7 matplotlib-3.10.0rc1 mpmath-1.4.0a2 pillow-11.0.0 platformdirs-4.3.6 pluggy-1.5.0 pooch-1.8.2 pyparsing-3.2.0 pytest-8.3.3 pytest-cov-6.0.0 pytest-xdist-3.6.1 python-dateutil-2.9.0.post0 requests-2.32.3 six-1.16.0 sortedcontainers-2.4.0 threadpoolctl-3.5.0 urllib3-2.2.3
Collecting scipy-openblas32==0.3.28.0.2 (from -r requirements/openblas.txt (line 1))
  Downloading scipy_openblas32-0.3.28.0.2-py3-none-manylinux_2_17_x86_64.manylinux2014_x86_64.whl.metadata (56 kB)
Downloading scipy_openblas32-0.3.28.0.2-py3-none-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (8.1 MB)
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 8.1/8.1 MB 128.6 MB/s eta 0:00:00
Installing collected packages: scipy-openblas32
Successfully installed scipy-openblas32-0.3.28.0.2
Looking in indexes: https://pypi.anaconda.org/scientific-python-nightly-wheels/simple
Requirement already satisfied: numpy in /opt/hostedtoolcache/Python/3.11.10/x64/lib/python3.11/site-packages (2.1.3)
Collecting numpy
  Downloading https://pypi.anaconda.org/scientific-python-nightly-wheels/simple/numpy/2.2.0.dev0/numpy-2.2.0.dev0-cp311-cp311-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (16.4 MB)
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 16.4/16.4 MB 100.0 MB/s eta 0:00:00
Installing collected packages: numpy
  Attempting uninstall: numpy
    Found existing installation: numpy 2.1.3
    Uninstalling numpy-2.1.3:
      Successfully uninstalled numpy-2.1.3
Successfully installed numpy-2.2.0.dev0
```

</details>

```
 C compiler for the host machine: ccache cc (gcc 11.4.0 "cc (Ubuntu 11.4.0-1ubuntu1~22.04) 11.4.0")
C linker for the host machine: cc ld.bfd 2.38
C++ compiler for the host machine: ccache c++ (gcc 11.4.0 "c++ (Ubuntu 11.4.0-1ubuntu1~22.04) 11.4.0")
C++ linker for the host machine: c++ ld.bfd 2.38
```

--- a/scipy/interpolate/_interpnd.pyx
+++ b/scipy/interpolate/_interpnd.pyx
@@ -336,30 +336,32 @@ class LinearNDInterpolator(NDInterpolatorBase):
         eps = 100 * DBL_EPSILON
         eps_broad = sqrt(DBL_EPSILON)
 
-        with nogil:
-            for i in range(xi.shape[0]):
+        # NOTE: a nogil block segfaults here with Python 3.10
+        # and 3.11 on x86_64 Ubuntu Linux with gcc 9.x and 11.x
+        # and therefore nogil was disabled to fix gh-21885
+        for i in range(xi.shape[0]):
 
-                # 1) Find the simplex
+            # 1) Find the simplex
 
-                isimplex = qhull._find_simplex(&info, c,
-                                               &xi[0,0] + i*ndim,
-                                               &start, eps, eps_broad)
+            isimplex = qhull._find_simplex(&info, c,
+                                           &xi[0,0] + i*ndim,
+                                           &start, eps, eps_broad)
 
-                # 2) Linear barycentric interpolation
+            # 2) Linear barycentric interpolation
 
-                if isimplex == -1:
-                    # don't extrapolate
-                    for k in range(nvalues):
-                        out[i,k] = fill_value
-                    continue
-
+            if isimplex == -1:
+                # don't extrapolate
                 for k in range(nvalues):
-                    out[i,k] = 0
+                    out[i,k] = fill_value
+                continue
 
-                for j in range(ndim+1):
-                    for k in range(nvalues):
-                        m = simplices[isimplex,j]
-                        out[i,k] = out[i,k] + c[j] * values[m,k]
+            for k in range(nvalues):
+                out[i,k] = 0
+
+            for j in range(ndim+1):
+                for k in range(nvalues):
+                    m = simplices[isimplex,j]
+                    out[i,k] = out[i,k] + c[j] * values[m,k]
 
         return out
 

